feat(dpg): basic/collection output model property

archerzz opened the pull request:
- make model field read-only if the model is output
- separate serialization constructor parameters from public constructors parameters
  - make serialization constructor parametes without validation, that fixes unnecessary null check in serialization constructors
  - change the initialization of public/serialization constructors accordingly
- make public constructor internal if there are collection properties
- fix duplicated fields initialization by excluding fiedls which are already initialized by parameters
- fix test cases by removing unnecessary null validation in serialization constructors
- add test cases
- refactor test cases

resolve #2462 #2470

# Description
<i>Add your description here!</i>

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

AlexanderSher commented on src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:
```diff
@@ -84,8 +85,12 @@ private static (IReadOnlyDictionary<FieldDeclaration, InputModelProperty> Fields
+                // * property is required and model is either input only or output only
                 var fieldModifiers = inputModelProperty.IsReadOnly || inputModelProperty.Type is InputDictionaryType or InputListType ||
-                    (inputModel.Usage == InputModelTypeUsage.Input && inputModelProperty.IsRequired)
+                    ((inputModel.Usage == InputModelTypeUsage.Input || inputModel.Usage == InputModelTypeUsage.Output) && inputModelProperty.IsRequired)
```
`inputModel.Usage is InputModelTypeUsage.Input or InputModelTypeUsage.Output` is a bit shorter

AlexanderSher commented on src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:
```diff
@@ -84,8 +85,12 @@ private static (IReadOnlyDictionary<FieldDeclaration, InputModelProperty> Fields
+                // property is read-only if:
+                // * model is read-only
+                // * property is a collection
+                // * property is required and model is either input only or output only
```
Let's formulate this in code:
```cs
var propertyIsCollection = inputModelProperty.Type is InputDictionaryType or InputListType;
var propertyIsRequiredInNonRoundtripModel = (inputModel.Usage is InputModelTypeUsage.Input or InputModelTypeUsage.Output) && inputModelProperty.IsRequired;
var propertyIsReadOnly = inputModelProperty.IsReadOnly || propertyIsCollection || propertyIsRequiredInNonRoundtripModel;
```

AlexanderSher commented on src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:
```diff
@@ -124,28 +129,59 @@ private static (ConstructorSignature PublicConstructor, ConstructorSignature Ser
+                var publicConstructorAccessbility = (usage == InputModelTypeUsage.Output ? MethodSignatureModifiers.Internal : MethodSignatureModifiers.Public);
+                if (serializationParameters.Any(p => TypeFactory.IsList(p.Type)) || !publicParameters.SequenceEqual(serializationParameters, new ParameterComparer()))
+                {
+                    serializationConstructor = new ConstructorSignature(name, $"Initializes a new instance of {name}", null, MethodSignatureModifiers.Internal, serializationParameters);
```
And that way you won't need a `ParameterComparer`

archerzz replied:
agree, since we have to convert `IList` to `IEnumerable` at the end, better to organize parameter conversion together.

archerzz commented:
changes since last:
- improve logical statement
- change the logic of removing validation of serialization constructor parameters
  - keep original parameters same for both public and serialization constructor
  - remove unnecessary custom comparer
  - remove validation of serialization constructor parameters when composing independent serialization constructor

ArcturusZhang reviewed (commented):
This may not be a this PR specific topic, since this PR only changes fields and properties.
Looks like we are using the `InputModelTypeUsage` to decide whether this class should be public or internal?
Could we consider that we generate every model as public, and introduce the `internalizer` and `remover` into DPG generation to prune the generated models? So that we could get rid of a lot of the if else clauses (maybe?). 
The internalizer and remover are using Roslyn to analyze things therefore no code model or cadl model issues involved.

archerzz commented:
change since last update:
- add test case for fixed/extensible enum properties of output model

archerzz commented:
> This may not be a this PR specific topic, since this PR only changes fields and properties. Looks like we are using the `InputModelTypeUsage` to decide whether this class should be public or internal? Could we consider that we generate every model as public, and introduce the `internalizer` and `remover` into DPG generation to prune the generated models? So that we could get rid of a lot of the if else clauses (maybe?). The internalizer and remover are using Roslyn to analyze things therefore no code model or cadl model issues involved.

@dapeng the input/output/roundtrip usage is calculated and passed in by CADL compiler. The getter/setter/accessibility/constructors will change according to the given usage. I think even if we use internalizer and remover, we still need to write codes to change those properties, so I'm sure if that will simplify the logic.

AlexanderSher reviewed (approved):
LGTM.
